3.13/specializer.md

```diff
@@ -0,0 +1,289 @@
+# Tier 2 Optimizer
+
+Author: Ken Jin, with implementation help by Jules Poon.
+
+This document explains how we plan to optimize CPython superblocks.
+See [execution engine](./engine.md) for more information about traces,
+superblocks and micro operations.
+
+## Overview
+
+After translation of Tier 1 specialized bytecode to Tier 2 micro
+operations, said micro operations need optimization to eventually
+outperform Tier 1.
+
+As with optimization in other languages, we require the following
+phases:
+- Analysis
+- Optimization
+- Code generation
+
+We will not cover code generation as it is kept simple for the tier
+2 interpreter.
+
+The optimized micro operations will then be passed to an exeuction
+engine, that is either the interpreter or the copy-and-patch JIT
+compiler.
+
+## Analysis
+
+We plan to generate majority of the analysis and optimization passes
+automatically, by defining bytecode semantics in the
+[interpreter DSL](../3.12/interpreter_definition.md). This will reduce
+the maintenance burden and lines of code that has to be checked into
+CPython. The general goal is that while certain instructions with
+special meaning to CPython have to be handwritten, the
+vast majority of other Python instructions's behavior in the
+interpreter can be automatically inferred from the DSL.
+
+To generate the intermediate representation, and perform
+some optimizations, we shall use an abstract interpreter over the
+Tier 2 uops. The abstract interpreter operates on and generates the IR
+below by operating on symbolic terms (as defined in the paper
+"SSA Translation Is an Abstract Interpretation" by Matthieu Lemerre
+in POPL 2023).
+
+CPython uops are then classified into 3 main types
+in the interpreter DSL:
+1. ``pure`` - like in functional languages,
+these instructions have no side effects to the user, they
+allow for greater optimizations.
+2. ``impure`` - everything that is not the above.
+3. ``guard`` - these instructions pass nodes through themselves.
+
+An extra modified - ``mandatory`` indicates that an instruction is
+always required and cannot be eliminated by static analysis. An example
+is state guards like checks for PEP 523.
+
+### Intermediate Representation (IR)
+
+The intended intermediate representation is a hybrid of assignments,
+plain instructions, and symbolic terms. That is, for the
+following Python code:
+
+```python
+x = a + b + c
+call(thing)
+```
+
+It will produce the following IR (assuming no guards are required):
+```
+x    := BINARY_OP_ADD_INT(a1, BINARY_OP_ADD_INT(b1, c1))
```

Comment:
Within symbolic terms, we have full code motion and easy common subexpression elimination (it's just a dict lookup). But it still maintains the required inherent ordering of a stack machine.

Micro-ops are not so mobile.

Reply:
It will only be worth it in the long run if we do more aggressive optimizations. Otherwise, I the old technique is probably faster and cheaper (memory-wise).